main.go

```diff
@@ -15,6 +15,108 @@
+	"github.com/stacklok/frizbee/pkg/replacer"
+	"github.com/stacklok/frizbee/pkg/utils/config"
+
+	"golang.org/x/oauth2"
+	"log"
+	"os"
+	"strings"
+)
+
 func main() {
-	// This is a placeholder for the main function
+	ctx := context.Background()
+	// Initialize the frizbee action
+	frizbeeAction, err := initAction(ctx)
+	if err != nil {
+		log.Fatalf("Error initializing action: %v", err)
+	}
+
+	// Run the frizbee action
+	err = frizbeeAction.Run(ctx)
+	if err != nil {
+		if errors.Is(err, action.ErrUnpinnedFound) {
+			log.Printf("Unpinned actions or container images found. Check the Frizbee Action logs for more information.")
+			os.Exit(1)
+		}
+		log.Fatalf("Error running action: %v", err)
+	}
+}
+
+// initAction initializes the frizbee action - reads the environment variables, creates the GitHub client, etc.
+func initAction(ctx context.Context) (*action.FrizbeeAction, error) {
+	// Get the GitHub token from the environment
+	token := os.Getenv("GITHUB_TOKEN")
+	if token == "" {
+		return nil, fmt.Errorf("GITHUB_TOKEN environment variable is not set")
+	}
+
+	// Create a new GitHub client
+	ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: token})
+	tc := oauth2.NewClient(ctx, ts)
+
+	// Get the GITHUB_REPOSITORY_OWNER
+	repoOwner := os.Getenv("GITHUB_REPOSITORY_OWNER")
+	if repoOwner == "" {
+		return nil, fmt.Errorf("GITHUB_REPOSITORY_OWNER environment variable is not set")
+	}
+
+	// Split the GITHUB_REPOSITORY environment variable to get repo name
+	repoFullName := os.Getenv("GITHUB_REPOSITORY")
+	if repoFullName == "" {
+		return nil, fmt.Errorf("GITHUB_REPOSITORY environment variable is not set")
+	}
+
+	// Clone the repository
+	fs, repo, err := cloneRepository(fmt.Sprintf("https://github.com/%s", repoFullName), repoOwner, token)
+	if err != nil {
+		return nil, fmt.Errorf("failed to clone repository: %w", err)
+	}
+
+	// Read the action settings from the environment and create the new frizbee replacers for actions and images
+	return &action.FrizbeeAction{
+		Client:            github.NewClient(tc),
+		Token:             token,
+		RepoOwner:         repoOwner,
+		RepoName:          strings.TrimPrefix(repoFullName, repoOwner+"/"),
+		ActionsPath:       os.Getenv("INPUT_ACTIONS"),
+		DockerfilesPath:   os.Getenv("INPUT_DOCKERFILES"),
+		KubernetesPath:    os.Getenv("INPUT_KUBERNETES"),
+		DockerComposePath: os.Getenv("INPUT_DOCKER_COMPOSE"),
+		OpenPR:            os.Getenv("INPUT_OPEN_PR") == "true",
+		FailOnUnpinned:    os.Getenv("INPUT_FAIL_ON_UNPINNED") == "true",
+		ActionsReplacer:   replacer.NewGitHubActionsReplacer(&config.Config{}).WithGitHubClientFromToken(token),
+		ImagesReplacer:    replacer.NewContainerImagesReplacer(&config.Config{}),
+		BFS:               fs,
+		Repo:              repo,
+	}, nil
+}
+
+// cloneRepository clones the repository and returns a billy.Filesystem interface to interact with it
+func cloneRepository(url, owner, accessToken string) (billy.Filesystem, *git.Repository, error) {
```

Comment:
I wonder if there is an alternative approach to cloning? This might take a long time if there are large blobs in the repo. 

I guess a lot of this is based on the approach, is this intercepting PR's or a workflow that is scheduled?

Reply:
The use case for this is when the actions runs from a scheduled run. We can do something else once we add support for checking the contents of a PR patch too.

Reply:
Intercepting the PRs would have the added benefit of not requiring as high privileges I think.

About cloning, I view it as acceptable for because the action is not typically run on infra we pay for :-) we can then optimize the clone with something like [git sparse checkout](https://git-scm.com/docs/git-sparse-checkout) since the action configuration already tells us which files or directories to look at.